BUG: Allow merging on Index vectors

gfyoung opened the pull request:
This behavior used to work in v0.19.0 and
is consistent with the documentation.

Closes #19038

gfyoung committed:
BUG: Allow merging on Index vectors

This behavior used to work in v0.19.0 and
is consistent with the documentation.

Closes gh-19038

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -814,12 +814,12 @@ def _get_merge_keys(self):
-            x, (np.ndarray, Series)) and len(x) == len(left)
-        is_rkey = lambda x: isinstance(
-            x, (np.ndarray, Series)) and len(x) == len(right)
+        is_lkey = lambda x: isinstance(x, list_types) and len(x) == len(left)
```
can you use ``is_list_like`` here?

gfyoung replied:
I thought about that, but passing in `list` or `tuple` is not accepted (later on, we check for the `dtype` attribute.

jreback replied:
maybe ``lambda x: is_list_like(x) and hasattr(x, 'dtype')``?

just trying to avoid hardcoding (actually you could add that as ``is_array_like`` in ``dtypes/common``, seems useful

gfyoung replied:
Yeah, `is_array_like` makes sense as an abstraction.  I'll add that.

jreback commented on pandas/tests/reshape/merge/test_merge.py:
```diff
@@ -1370,6 +1370,22 @@ def f():
             household.join(log_return, how='outer')
         pytest.raises(NotImplementedError, f)
 
+    def test_merge_datetime_index(self):
```
can you parameterize on ``ndarray, Series, Index`` (for the df.index.year arg)

gfyoung replied:
Done.

gfyoung committed:
ENH: Add is_array_like method

Used for abstracting checks in DataFrame.merge,
but the function itself can be quite useful.

jreback commented on pandas/core/dtypes/inference.py:
```diff
@@ -267,6 +267,26 @@ def is_list_like(obj):
+    -------
+    is_array_like : bool
+        Whether `obj` has array-like properties.
+    """
```
can you show Index, Series, ndarray (and a scalar) as examples

jreback replied:
as a followup we could replace code in various places to use this

gfyoung replied:
Done.

gfyoung commented:
@jreback : Comments have been addressed, and all is green.  PTAL.

jreback commented:
very nice @gfyoung thanks!
